pyomo/contrib/solver/sol_reader.py

```diff
@@ -0,0 +1,207 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+
+from typing import Tuple, Dict, Any, List
+import io
+
+from pyomo.common.errors import DeveloperError, PyomoException
+from pyomo.repn.plugins.nl_writer import NLWriterInfo
+from pyomo.contrib.solver.results import Results, SolutionStatus, TerminationCondition
+
+
+class SolFileData:
+    def __init__(self) -> None:
+        self.primals: List[float] = list()
+        self.duals: List[float] = list()
+        self.var_suffixes: Dict[str, Dict[int, Any]] = dict()
+        self.con_suffixes: Dict[str, Dict[Any]] = dict()
+        self.obj_suffixes: Dict[str, Dict[int, Any]] = dict()
+        self.problem_suffixes: Dict[str, List[Any]] = dict()
+        self.other: List(str) = list()
+
+
+def parse_sol_file(
+    sol_file: io.TextIOBase, nl_info: NLWriterInfo, result: Results
+) -> Tuple[Results, SolFileData]:
```

Comment:
I think we should re-think this API and not pass the results in / out of this function.  I would argue for this to be a _parser_ that returns the parsed data, and populating a results object is the responsibility of the caller.  This would also remove the somewhat confusing passing in / out of results objects.

Reply:
I agree. It's "rough" but not "as rough as it used to be" right meow.